File: frontend/src/lib/conditions.js
/**
 * D&D 5.5e (2024) conditions — Rules Glossary.
 * Conditions don't stack with themselves; Exhaustion is the exception (levels 1–6).
 */

export const CONDITION_OPTIONS = [
  "Blinded",
  "Charmed",
  "Deafened",
  "Exhaustion",
  "Frightened",
  "Grappled",
  "Incapacitated",
  "Invisible",
  "Paralyzed",
  "Petrified",
  "Poisoned",
  "Prone",
  "Restrained",
  "Stunned",
  "Unconscious",
];

/** Short rules summary for UI tooltips (2024 glossary). */
export const CONDITION_HINTS = {
  Blinded: "Can't see; attacks have Disadvantage; attacks against you have Advantage.",
  Charmed: "Can't attack the charmer; charmer has Advantage on social checks against you.",
  Deafened: "Can't hear; automatically fail hearing-based checks.",
  Exhaustion: "Cumulative levels 1–6. −2 per level to D20 Tests; −5 ft. Speed per level. Die at 6.",
  Frightened: "Disadvantage on checks/attacks while source is in line of sight; can't willingly move closer.",
  Grappled: "Speed 0; ends if grappler is Incapacitated or you escape.",
  Incapacitated: "Can't take actions, Bonus Actions, or Reactions; can't speak (except faltering if Stunned).",
  Invisible: "Heavily obscured for hiding; attacks against you have Disadvantage; your attacks have Advantage.",
  Paralyzed: "Incapacitated; Speed 0; auto-fail Str/Dex saves; attacks against you have Advantage; melee hits within 5 ft. are crits.",
  Petrified: "Incapacitated; Speed 0; Resistant to all damage; Immune to Poisoned.",
  Poisoned: "Disadvantage on attack rolls and ability checks.",
  Prone: "Disadvantage on attacks; melee attacks against you have Advantage, ranged have Disadvantage.",
  Restrained: "Speed 0; Disadvantage on attacks and Dex saves; attacks against you have Advantage.",
  Stunned: "Incapacitated; Speed 0; auto-fail Str/Dex saves; attacks against you have Advantage.",
  Unconscious: "Incapacitated and Prone; Speed 0; unaware; auto-fail Str/Dex saves; melee hits within 5 ft. are crits.",
};

const INCAPACITATING = new Set(["Paralyzed", "Stunned", "Unconscious", "Petrified"]);

export function normalizeConditions(value) {
  if (Array.isArray(value)) {
    return value.map((entry) => String(entry).trim()).filter(Boolean);
  }
  if (value == null || !String(value).trim()) return [];
  return String(value)
    .split(",")
    .map((part) => part.trim())
    .filter(Boolean);
}

export function getExhaustionLevel(conditions) {
  const list = normalizeConditions(conditions);
  const entry = list.find((c) => /^exhaustion\b/i.test(c));
  if (!entry) return 0;
  const match = entry.match(/^exhaustion\s*(\d+)?/i);
  if (!match) return 1;
  const level = match[1] ? parseInt(match[1], 10) : 1;
  return Number.isFinite(level) && level > 0 ? level : 1;
}

export function formatExhaustion(level) {
  return `Exhaustion ${level}`;
}

export function hasCondition(conditions, name) {
  const list = normalizeConditions(conditions);
  const lower = name.toLowerCase();
  if (lower === "exhaustion") return getExhaustionLevel(list) > 0;
  return list.some((entry) => entry.toLowerCase() === lower);
}

function withoutExhaustion(conditions) {
  return normalizeConditions(conditions).filter((entry) => !/^exhaustion\b/i.test(entry));
}

/**
 * @returns {{ ok: true, conditions: string[] } | { ok: false, reason: string }}
 */
export function addCondition(conditions, conditionName, { dryRun = false } = {}) {
  const list = normalizeConditions(conditions);
  const name = String(conditionName || "").trim();
  if (!name) return { ok: false, reason: "Choose a condition." };

  if (name === "Exhaustion") {
    const level = getExhaustionLevel(list);
    if (level >= 6) {
      return { ok: false, reason: "Exhaustion is already level 6 — the creature dies." };
    }
    const next = [...withoutExhaustion(list), formatExhaustion(level + 1)];
    return dryRun ? { ok: true, conditions: next } : { ok: true, conditions: next };
  }

  if (hasCondition(list, name)) {
    return {
      ok: false,
      reason: `${name} is already applied. Conditions don't stack with themselves (5.5e Rules Glossary).`,
    };
  }

  if (name === "Incapacitated" && impliesIncapacitated(list)) {
    return {
      ok: false,
      reason:
        "Incapacitated is already in effect via Paralyzed, Stunned, Unconscious, or Petrified.",
    };
  }

  if (name === "Prone" && hasCondition(list, "Unconscious")) {
    return {
      ok: false,
      reason: "Prone is already applied while a creature has the Unconscious condition.",
    };
  }

  if (name === "Poisoned" && hasCondition(list, "Petrified")) {
    return {
      ok: false,
      reason: "Petrified creatures have Immunity to the Poisoned condition.",
    };
  }

  const next = [...list, name];
  return dryRun ? { ok: true, conditions: next } : { ok: true, conditions: next };
}

export function removeCondition(conditions, conditionName) {
  const list = normalizeConditions(conditions);
  const lower = String(conditionName).toLowerCase();
  if (lower.startsWith("exhaustion")) {
    return withoutExhaustion(list);
  }
  return list.filter((entry) => entry.toLowerCase() !== lower);
}

export function canAddCondition(conditions, conditionName) {
  return addCondition(conditions, conditionName, { dryRun: true }).ok;
}

export function impliesIncapacitated(conditions) {
  const list = normalizeConditions(conditions);
  if (hasCondition(list, "Incapacitated")) return true;
  return list.some((entry) => {
    const base = entry.split(/\s+/)[0];
    return INCAPACITATING.has(base);
  });
}

export function formatConditionsList(conditions) {
  return normalizeConditions(conditions).join(", ");
}

File: backend/app/services/conditions.py
"""D&D 5.5e (2024) condition rules for encounter combatants."""

from __future__ import annotations

import re

CONDITION_OPTIONS = [
    "Blinded",
    "Charmed",
    "Deafened",
    "Exhaustion",
    "Frightened",
    "Grappled",
    "Incapacitated",
    "Invisible",
    "Paralyzed",
    "Petrified",
    "Poisoned",
    "Prone",
    "Restrained",
    "Stunned",
    "Unconscious",
]

_INCAPACITATING = frozenset({"Paralyzed", "Stunned", "Unconscious", "Petrified"})


def normalize_conditions(value: list[str] | str | None) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(entry).strip() for entry in value if str(entry).strip()]
    text = str(value).strip()
    if not text:
        return []
    return [part.strip() for part in text.split(",") if part.strip()]


def get_exhaustion_level(conditions: list[str] | str | None) -> int:
    for entry in normalize_conditions(conditions):
        if not re.match(r"^exhaustion\b", entry, re.IGNORECASE):
            continue
        match = re.match(r"^exhaustion\s*(\d+)?", entry, re.IGNORECASE)
        if not match:
            return 1
        level = int(match.group(1)) if match.group(1) else 1
        return level if level > 0 else 1
    return 0


def _format_exhaustion(level: int) -> str:
    return f"Exhaustion {level}"


def has_condition(conditions: list[str] | str | None, name: str) -> bool:
    lower = name.lower()
    if lower == "exhaustion":
        return get_exhaustion_level(conditions) > 0
    return any(entry.lower() == lower for entry in normalize_conditions(conditions))


def _implies_incapacitated(conditions: list[str] | str | None) -> bool:
    for entry in normalize_conditions(conditions):
        base = entry.split()[0] if entry.split() else entry
        if base in _INCAPACITATING:
            return True
    return False


def _without_exhaustion(conditions: list[str]) -> list[str]:
    return [entry for entry in conditions if not re.match(r"^exhaustion\b", entry, re.IGNORECASE)]


def validate_add_condition(
    conditions: list[str] | str | None, condition_name: str
) -> tuple[bool, list[str], str | None]:
    """Return (ok, next_list, error_message)."""
    listed = normalize_conditions(conditions)
    name = condition_name.strip()
    if not name:
        return False, listed, "Choose a condition."

    if name == "Exhaustion":
        level = get_exhaustion_level(listed)
        if level >= 6:
            return False, listed, "Exhaustion is already level 6 — the creature dies."
        next_list = [*_without_exhaustion(listed), _format_exhaustion(level + 1)]
        return True, next_list, None

    if has_condition(listed, name):
        return (
            False,
            listed,
            f"{name} is already applied. Conditions don't stack with themselves (5.5e Rules Glossary).",
        )

    if name == "Incapacitated" and _implies_incapacitated(listed):
        return (
            False,
            listed,
            "Incapacitated is already in effect via Paralyzed, Stunned, Unconscious, or Petrified.",
        )

    if name == "Prone" and has_condition(listed, "Unconscious"):
        return (
            False,
            listed,
            "Prone is already applied while a creature has the Unconscious condition.",
        )

    if name == "Poisoned" and has_condition(listed, "Petrified"):
        return (
            False,
            listed,
            "Petrified creatures have Immunity to the Poisoned condition.",
        )

    return True, [*listed, name], None


def sanitize_conditions_list(conditions: list[str] | str | None) -> list[str]:
    """Normalize legacy strings and drop redundant implied conditions."""
    listed = normalize_conditions(conditions)
    out: list[str] = []
    exhaustion_level = 0

    for entry in listed:
        if re.match(r"^exhaustion\b", entry, re.IGNORECASE):
            match = re.match(r"^exhaustion\s*(\d+)?", entry, re.IGNORECASE)
            level = int(match.group(1)) if match and match.group(1) else 1
            exhaustion_level = max(exhaustion_level, min(6, max(1, level)))
            continue
        base = entry.split()[0] if entry.split() else entry
        if base in CONDITION_OPTIONS and base not in out:
            out.append(base)

    if exhaustion_level:
        out.append(_format_exhaustion(exhaustion_level))

    if _implies_incapacitated(out) and "Incapacitated" in out:
        out = [entry for entry in out if entry != "Incapacitated"]
    if has_condition(out, "Unconscious") and "Prone" in out:
        out = [entry for entry in out if entry != "Prone"]
    if has_condition(out, "Petrified") and "Poisoned" in out:
        out = [entry for entry in out if entry != "Poisoned"]

    return out


def assert_conditions_valid(conditions: list[str] | str | None) -> list[str]:
    """Raise ValueError when a condition list breaks 5.5e stacking rules."""
    listed = sanitize_conditions_list(conditions)
    bases = [entry for entry in listed if not re.match(r"^exhaustion\b", entry, re.IGNORECASE)]
    if len(bases) != len(set(bases)):
        raise ValueError("Each condition can only be applied once (Exhaustion excepted).")

    exhaustion_level = get_exhaustion_level(listed)
    if exhaustion_level > 6:
        raise ValueError("Exhaustion cannot exceed level 6.")

    for base in bases:
        if base not in CONDITION_OPTIONS:
            raise ValueError(f"Unknown condition: {base}")

    if "Incapacitated" in bases and _implies_incapacitated(bases):
        raise ValueError(
            "Incapacitated is already in effect via Paralyzed, Stunned, Unconscious, or Petrified."
        )
    if "Prone" in bases and has_condition(bases, "Unconscious"):
        raise ValueError("Prone is already applied while a creature has the Unconscious condition.")
    if "Poisoned" in bases and has_condition(bases, "Petrified"):
        raise ValueError("Petrified creatures have Immunity to the Poisoned condition.")

    return listed
